libc-bottom-half/sources/chdir.c

```diff
@@ -71,6 +72,31 @@ int chdir_legacy(const char *path)
 }
 
 static const char *make_absolute(const char *path) {
+    // if the libc has not synced with the runtime yet, call into the runtime to get the cwd
+    if (!__wasilibc_cwd_is_synced) {
+        char cwd[256];
+        if (getcwd(cwd, sizeof(cwd)) != NULL) {
+            char *new_cwd = malloc(strlen(cwd) + 1);
+            if (new_cwd != NULL) {
+                strcpy(new_cwd, cwd);
+            }
+
+            __wasilibc_cwd_lock();
+            char *prev_cwd = __wasilibc_cwd;
+            __wasilibc_cwd = new_cwd;
+            __wasilibc_cwd_unlock();
+
+            if (__wasilibc_cwd_mallocd)
+                free(prev_cwd);
```

Comment:
This doesn't look thread-safe.

If a different thread had acquired the old pointer, but not used it yet, this could lead to invalid access.

Eg, just below is a ` int need_slash = __wasilibc_cwd[cwd_len - 1] == '/' ? 0 : 1;` that happens outside of holding the lock.

Since we don't expect this change to happen often from the host side, maybe we just want to leak the value?

Reply:
> Since we don't expect this change to happen often

Correction: we don't expect it to happen at all from the host side, there's nothing (that I know of at least) that changes the CWD on the host side.

However, you raise a valid point; I also believe the value should be leaked. The question is what happens when CWD changes from the guest side and how that interacts with this change.

Reply:
In theory being able to change the CWD from the host side would be quite nice, eg in the context of an interactive editor or similar, but I don't think it's something we need to account for.